Framework/DataHandling/src/LoadDNSEvent.cpp

```diff
@@ -0,0 +1,525 @@
+    // uint64_t chopperTimestamp = 0;
+    uint64_t oversizedChanelIndexCounter = 0;
+    uint64_t oversizedPosCounter = 0;
+    uint64_t i = 0;
+    const auto wsIndex = j;
+    auto &eventList = finalEventAccumulator.neutronEvents[j];
+    if (eventList.size() != 0) {
+      std::sort(eventList.begin(), eventList.end(), [](auto l, auto r) { return l.timestamp < r.timestamp; });
+    }
+
+    auto chopperIt = finalEventAccumulator.triggerEvents.cbegin();
+
+    auto &spectrum = eventWS->getSpectrum(wsIndex);
+    // PARALLEL_START_INTERUPT_REGION
+
+    for (const auto &event : eventList) {
+      i++;
+      if (i % EVENTS_PER_PROGRESS == 0) {
+        progress.report();
+        if (this->getCancel()) {
+          throw CancelException();
+        }
+      }
+
+      chopperIt =
+          std::lower_bound(finalEventAccumulator.triggerEvents.cbegin(), finalEventAccumulator.triggerEvents.cend(),
+                           event.timestamp, [](auto l, auto r) { return l.timestamp > r; });
+      const uint64_t chopperTimestamp = chopperIt != finalEventAccumulator.triggerEvents.cend()
+                                            ? chopperIt->timestamp
+                                            : 0; // before first chopper trigger
+      if ((chopperTimestamp == 0) && discardPreChopperEvents) {
+        // throw away events before first chopper trigger
+        continue;
+      }
+      spectrum.addEventQuickly(Types::Event::TofEvent(double(event.timestamp - chopperTimestamp) / 10.0));
+    }
+
+    // PARALLEL_END_INTERUPT_REGION
+    oversizedChanelIndexCounterA += oversizedChanelIndexCounter;
+    oversizedPosCounterA += oversizedPosCounter;
+    // PARALLEL_CHECK_INTERUPT_REGION
+  }
+
+  if (oversizedChanelIndexCounterA > 0) {
+    g_log.warning() << "Bad chanel indices: " << oversizedChanelIndexCounterA << std::endl;
+  }
+  if (oversizedPosCounterA > 0) {
+    g_log.warning() << "Bad position values: " << oversizedPosCounterA << std::endl;
+  }
+}
+
+namespace {
+template <typename Iterable>
+constexpr std::map<std::remove_reference_t<decltype(*std::declval<decltype(std::declval<Iterable>().cbegin())>())>,
+                   size_t>
+buildSkipTable(const Iterable &iterable) {
+  std::map<std::remove_reference_t<decltype(*std::declval<decltype(std::declval<Iterable>().cbegin())>())>, size_t>
+      skipTable;
+
+  size_t i = iterable.size();
+  for (auto c : iterable) {
+    i--;
+    auto &cPos = skipTable[c];
+    cPos = cPos == 0 ? iterable.size() : cPos;
+    cPos = i == 0 ? static_cast<size_t>(cPos) : i;
+  }
+  return skipTable;
+}
+
+template <typename Iterable> const std::vector<uint8_t> buildSkipTable2(const Iterable &iterable) {
+  std::vector<uint8_t> skipTable(256, static_cast<uint8_t>(iterable.size()));
+
+  size_t i = iterable.size();
+  for (auto c : iterable) {
+    i--;
+    auto &cPos = skipTable[c];
+    cPos = i == 0 ? cPos : static_cast<uint8_t>(i);
+  }
+  return skipTable;
+}
```

Comment:
I am not sure, how to replace this. The code is not simply doing string search.  It is incrementally loading the data file, in the steps given by Boyer Moore. If I want to do pure string search, I would have to load an unknown large part of the datafile.

Reply:
Ah I think we can use it at least to create the table.

Reply:
Have we measured the time in doing this search on-disk vs loading all data in then parsing / filtering in-memory?
I worry this is a case of premature optimisation, if it's a significant difference I'm happy to take it. But I worry we're trading a couple of seconds of loading time (max) for significantly harder to maintain code

Reply:
Hi David,
I have not originally written this, so I have not done any test if it was necessary. Was originally written by Joachim Coenen. I just fixed everything which was broken.

Loading everything in memory is not really an option I think, we are talking about hunderts of Mb per data file file and Gbs in total.
I would rather try, to do bytewise incrementally load, and check every time if the string matches, with boost or normal string  search. This would only be a marginal change to the code and would increase the checks by the time you would save by Boyer Moore, but could well be irrelevant for the total loading time. 
String matching should be very fast. For the header, it is only 2 lines to check at DNS, so the speed difference should be marginal. For the blocks its 1500byte, and around 8 blocks for file splitting. 
Compared to Billions of events which are processed, this should not matter.

Just made some quick tests, setting the skip table to 1 byte for everything.
There is no visible difference in speed, it is hidden in the fluctuations, for 1 Billion of counts less than 0.02 s.

Reply:
With modern optimising compilers I wouldn't be surprised if there's some tricks being inlined that causes the naive approach to be pretty speedy.

Thanks for testing that, I was on the fence about asking since you've already done lots of work for this PR. But good to see it all ultimately worked out :+1: